### ISSUE TITLE
Add option for automatic pricing with average price

### ISSUE THREAD

**Issue by Snackhole** [ArkadiusTradeTools/arkadius-trade-tools#39]
Used to use Master Merchant before it became too onerous to keep its data up to date.  The one feature I missed from it when switching to ATT was that Master Merchant would automatically price items for sale in the guild store using the average price, which I found very convenient (I'm lazy and didn't want to manually type or do research on the price, and so on).

I missed this so much that I tracked down how Master Merchant was doing it and created a dependent addon to adapt the function to ATT, composed of the following files:

###### ATTPricer.txt

 ```
## Title: ATT Pricer
## Author: Snackhole
## Version: 1.7.2
## AddOnVersion: 1
## APIVersion: 100032
## DependsOn: ArkadiusTradeTools ArkadiusTradeToolsSales

ATTPricer.lua
```

###### ATTPricer.lua:

```
ATTPricer = {}

ATTPricer.name = "ATTPricer"
local OriginalSetupPendingPost
local SECONDS_IN_DAY = 60 * 60 * 24

function ATTPricer.SetupPendingPost(self)
    OriginalSetupPendingPost(self)
    if (self.pendingItemSlot) then
        local itemLink = GetItemLink(BAG_BACKPACK, self.pendingItemSlot)
        local _, stackCount, _ = GetItemInfo(BAG_BACKPACK, self.pendingItemSlot)
        local days = ArkadiusTradeTools.Modules.Sales.TradingHouse:GetCalcDays()
        local AveragePrice = ArkadiusTradeTools.Modules.Sales:GetAveragePricePerItem(itemLink, GetTimeStamp() - SECONDS_IN_DAY * days)
		
        if AveragePrice then
            self:SetPendingPostPrice(math.floor(AveragePrice * stackCount))
        end
    end
end

if TRADING_HOUSE then
    OriginalSetupPendingPost = TRADING_HOUSE.SetupPendingPost
    TRADING_HOUSE.SetupPendingPost = ATTPricer.SetupPendingPost
end
```

It seems unlikely that every user of the addon would want this behavior, so I figure if it _were_ to be implemented into ATT it should be as an option, not just default and immutable behavior.  I can also say that I have used this exact code personally without issue for months now and it has no impact on my game's performance or UI beyond the intended functionality.  That said I do not have a lot of experience with Lua or ESO addons specifically, so I'm sure this code has lots of room for improvement.

If this is something that has been considered already and rejected for whatever reason, at the very least I figure this issue gives publicly posted minimal working code for anyone who wants to just copy and paste into their own addons folder.

Edit:  Apparently copying into GitHub's editor messed up the whitespace in the Lua.

**Comment by timothymclane** [comment]
Thanks for opening the issue and putting together a working example. :)

The master branch currently has support for auto pricing with the vanilla UI, and has a button for auto pricing when using AGS. There's a setting to toggle this functionality. I'm not sure if this does exactly what you want, but you can pull down latest from master and see if it works like you'd expect. If not, we can see what we can do to provide the functionality you're looking for.

I haven't been able to cut a release due to some familial events in life, but I'm hoping to push everything to ESOUI sometime within the next week, barring any real-world emergencies.

**Comment by Snackhole** [comment]
Sure, I'll try to take a look at it tonight!  If I understand your description correctly, it seems exactly like what I'd do/want!

**Comment by Snackhole** [comment]
Hm, not sure if I'm doing something wrong...  Here's what I did:

1. Downloaded master branch as a .zip file.
2. Deleted files that didn't seem to be directly related to the addon's functioning (basically everything except the 21 folders starting with "ArkadiusTradeTools").
3. Replaced my copies of those folders in the AddOns folder.
4. Started the game.
5. Turned on the setting for autopricing.
6. Went to the guild store and started listing whatever I had on hand to see if it worked.

It seemed to be merely tripling the vendor price?  So a stack of lockpicks was priced at 0, while a stack of filled soul gems was priced at exactly 18,000, neither of which lines up with their average prices in my guild.  I'm not sure if that's vanilla behavior or not.  I also got an error on opening the guild store, but I was too rushed to capture it since I did this right before eating dinner.  I can report back with the error later if you like.

**Comment by timothymclane** [comment]
Are you using Awesome Guild Store or vanilla?

**Comment by Snackhole** [comment]
Vanilla!

**Comment by timothymclane** [comment]
OK. Then I'll need that UI Error message.

**Comment by Snackhole** [comment]
Same error whether the setting is enabled or not:

```
user:/AddOns/ArkadiusTradeToolsSales/ArkadiusTradeToolsSalesTradingHouse.lua:209: table index is nil
stack traceback:
user:/AddOns/ArkadiusTradeToolsSales/ArkadiusTradeToolsSalesTradingHouse.lua:209: in function 'ZO_ScrollList_Commit_Hook'
|caaaaaa<Locals> list = ud, scrollData = [table:1]{}, averagePrices = [table:2]{}, days = 10, i = 1, itemType = 0 </Locals>|r
EsoUI/Libraries/Utility/ZO_Hook.lua:18: in function 'ZO_ScrollList_Commit'
EsoUI/Ingame/TradingHouse/Keyboard/TradingHouse_Keyboard.lua:865: in function 'ZO_TradingHouseManager:RebuildSearchResultsPage'
|caaaaaa<Locals> self = [table:3]{initialized = T, pendingSaleIsValid = F, currentMode = 1, currentDisplayName = "@Snackhole", fireCallbackDepth = 0}, isInitialResults = T, list = ud, scrollData = [table:1], showingGuildSpecificItems = T, numItemsOnPage = 1 </Locals>|r
EsoUI/Ingame/TradingHouse/Keyboard/TradingHouse_Keyboard.lua:1019: in function 'ZO_TradingHouseManager:UpdateForGuildChange'
|caaaaaa<Locals> self = [table:3], guildId = 303181, canSell = T, IS_INITIAL_RESULTS = T </Locals>|r
EsoUI/Ingame/TradingHouse/Keyboard/TradingHouse_Keyboard.lua:987: in function 'ZO_TradingHouseManager:UpdateStatus'
|caaaaaa<Locals> self = [table:3] </Locals>|r
EsoUI/Ingame/TradingHouse/Keyboard/TradingHouse_Keyboard.lua:57: in function 'OnUpdateStatus'
EsoUI/Ingame/TradingHouse/Keyboard/TradingHouse_Keyboard.lua:42: in function '(anonymous)'
```

**Comment by timothymclane** [comment]
That doesn't look possible if you're on the latest version. I'd recommend exiting ESO, deleting or moving your ATT addon folders out of your addons directory (keep your saved vars), and then copy over the ones from the zip you downloaded earlier. Then let me know if it's still happening.

I don't have a PC handy where I can investigate, but I should be able to tomorrow.

**Comment by Snackhole** [comment]
That's exactly what I did; I moved my ATT addon folders out and copied the ones from the zip, both the first time I tested and the time I got the error copied down.

**Comment by timothymclane** [comment]
I believe you. The error just looks like it's not possible on that line, so it made me think you might've had an old file that didn't get overwritten correctly. Just trying to cover our bases.

I'll have to dig into this in the next couple of days. I'll comment here when I learn more.

**Comment by Snackhole** [comment]
Sounds good!  Sorry, I didn't expect this to be such a can of worms when I posted it!

**Comment by timothymclane** [comment]
I've been able to replicate your issue. I'm not sure what's going on yet, but I'll let you know what I find.

**Comment by timothymclane** [comment]
@Snackhole, I believe I've fixed the issue. Go ahead and try the current master branch when you get a chance.

**Comment by Snackhole** [comment]
Okay, no error this time, but the other issue remains.  Whether I enable the automatic pricing setting or not, the behavior is identical:  It sets the price automatically to something _other_ than guild price times stack size.  The three items I've tested:

* 200 lockpicks; 0 vendor price, 0 automatic price in guild store, 24.19 guild price per item, 4,838 expected price
* 200 filled soul gems; 6,000 vendor price, 18,000 automatic price in guild store, 33.97 guild price per item, 6,794 expected price
* 13 melon-baked parmesan pork; 65 vendor price, 195 automatic price in guild store, 26.8 guild price per item, 348 expected price

I also disabled the addon entirely and got the same results, so I think it's just vanilla pricing behavior.  I don't have any other addons that do anything relevant to the guild store, I think:

* Advanced Filters - Updated
* Chat Window Manager
* Dolgubon's Lazy Writ Crafter
* Dungeon Champions
* Harven's Improved Skills Window
* Harven's Stack Split Slider
* LoreBooks
* Lost Treasure
* SkyShards
* SpentSkillPoints
* Votan's Keybinder
* Votan's Minimap

**Comment by Snackhole** [comment]
To be absolutely clear, there's no button to click if I'm not using AGS, right?  It should just automatically set the price to the stack size times the average guild price when I select an item to sell?

**Comment by timothymclane** [comment]
Yes, it should set the price automatically.

I didn't test that functionality while fixing the error as I wasn't sure the guild store integration was actually working with the error being thrown.

I'll take another look in a few hours and see if I can track down what's happening.

**Comment by timothymclane** [comment]
This appears to be working on my end, so I'm not sure what's happening on yours.

Could you share screen snips or a video of the item tooltip and what it looks like when you list it in store? Your ArkadiusTradeToolsSales.lua saved variables file would also be helpful.

**Comment by Snackhole** [comment]
I've attached screenshots as well as .zip files of the two ATT installations I'm swapping between for testing, the one I had before (what's uploaded to ESOUI plus my custom addon for pricing) and the one I downloaded from the master branch after your update that fixed the error.  And here is my ArkadiusTradeToolsSales.lua file:

```
ArkadiusTradeToolsSalesData =
{
    ["settings"] = 
    {
        ["tradingHouse"] = 
        {
            ["enabled"] = true,
            ["calcDays"] = 10,
            ["enableAutoPricing"] = true,
            ["defaultDealLevel"] = 2,
        },
        ["tooltips"] = 
        {
            ["days"] = 10,
            ["enabled"] = true,
            ["graphEnabled"] = false,
            ["craftingEnabled"] = true,
        },
        ["filters"] = 
        {
            ["itemName"] = true,
            ["useSubStrings"] = true,
            ["price"] = false,
            ["timeStamp"] = false,
            ["guildName"] = true,
            ["buyerName"] = false,
            ["timeSelection"] = 4,
            ["sellerName"] = true,
        },
        ["guilds"] = 
        {
            ["The Ninja Squirrels"] = 
            {
                ["keepSalesForDays"] = 30,
            },
        },
        ["hiddenHeaders"] = 
        {
        },
        ["guildRoster"] = 
        {
            ["timeSelectionIndex"] = 1,
            ["enabled"] = true,
        },
        ["inventories"] = 
        {
            ["enabled"] = false,
        },
    },
}
```

[arkadius-trade-tools-master.zip](https://github.com/ArkadiusTradeTools/arkadius-trade-tools/files/5929987/arkadius-trade-tools-master.zip)
[Working ATT Installation.zip](https://github.com/ArkadiusTradeTools/arkadius-trade-tools/files/5929988/Working.ATT.Installation.zip)
![01 Setting](https://user-images.githubusercontent.com/37544796/106987854-f7019080-6733-11eb-91e7-68b8dca39b5d.png)
![02 Lockpick Tooltip](https://user-images.githubusercontent.com/37544796/106987856-f79a2700-6733-11eb-818a-519317c48742.png)
![03 Lockpick Pricing](https://user-images.githubusercontent.com/37544796/106987857-f8cb5400-6733-11eb-80a4-b9aeb2074ae2.png)
![04 Pork Tooltip](https://user-images.githubusercontent.com/37544796/106987858-f8cb5400-6733-11eb-92b3-1e2acbdd7f0a.png)
![05 Pork Pricing](https://user-images.githubusercontent.com/37544796/106987859-f963ea80-6733-11eb-9e45-4378b73eee55.png)
![06 Soul Gem Tooltip](https://user-images.githubusercontent.com/37544796/106987862-f963ea80-6733-11eb-8a21-bc98f1d95f96.png)
![07 Soul Gem Pricing](https://user-images.githubusercontent.com/37544796/106987864-f9fc8100-6733-11eb-94a5-9c1c0734aaf2.png)

**Comment by timothymclane** [comment]
Thanks for your patience and thorough data. It's been very helpful. 😀

I'm not entirely sure, but this appears to be a race condition. I've been able to reproduce it, but it only happens when I have very few addons enabled, which makes me think some other addons are slowing the frontend sandbox down, causing it to run after the ZOS code.

I'll play around with it and see where I can reliably slide it in without getting too gross. I'll probably have something you can try out tomorrow.

**Comment by timothymclane** [comment]
Sorry about the delay in getting back to you. Yesterday got away from me.

I've tested what's in master on my machine with all other addons disabled. It should fix the issue on your end. Let me know if you run into any more problems.

**Comment by Snackhole** [comment]
Happy to report it now works exactly as described on my end!

I noticed that it seems to round to the nearest integer instead of always to the floor like my own version, but that's a very minor difference and something I assume is a simple matter of preference.  Other than that it seems indistinguishable from what I made for myself (though I assume much more compatible with other addons).

Thanks for all the time you took to work on this.  I didn't expect to have brought so much hassle into your week when I posted this!
